### PR TITLE
Proxy support for cheatcodes

### LINKED ISSUE #279: Make `roll`, `prank` and `warp`  work in recursive contract calls and proxies
### To what component is your bug related to?

Forge

### Feature Request

We should intercept syscalls and make it use our cheatable entrypoint
I have a test ready - if anybody wants the code for TDD please contact @Arcticae

### PR DESCRIPTION
**Pull request by MaksymilianDemitraszek**
Closes #279 

## Introduced changes
- Cheatcodes work for nested calls

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code

### CODE REVIEW

File: starknet-foundry/crates/cheatable-starknet/src/rpc.rs [hunk 1]
```diff
@@ -474,9 +482,76 @@ impl CheatableSyscallHandler<'_> {
             response_wrapper.write(vm, &mut self.syscall_handler.syscall_ptr)?;
             return Ok(());
         }
+
+        if let SyscallSelector::CallContract = selector {
```
Comment (piotmag769) [inline starknet-foundry/crates/cheatable-starknet/src/rpc.rs]:
```suggestion
        if selector == SyscallSelector::CallContract {
```

Comment (piotmag769) [reply starknet-foundry/crates/cheatable-starknet/src/rpc.rs]: Could we maybe add `else if`?  I guess this only makes sense if address is not cheated, it is less confusing this way and we do not have to look up a couple hundreds lines of code

File: starknet-foundry/crates/cheatable-starknet/src/rpc.rs [hunk 2]
```diff
@@ -474,9 +482,76 @@ impl CheatableSyscallHandler<'_> {
...
     }
 
+    fn execute_syscall<Request, Response, ExecuteCallback>(
+        &mut self,
+        vm: &mut VirtualMachine,
+        execute_callback: ExecuteCallback,
+        base_gas_cost: u64,
+    ) -> HintExecutionResult
+    where
+        Request: SyscallRequest + std::fmt::Debug,
```
Comment (piotmag769) [inline starknet-foundry/crates/cheatable-starknet/src/rpc.rs]: Why do we need Debug trait here?

File: crates/cheatable-starknet/src/rpc.rs [hunk 3]
```diff
@@ -489,6 +564,75 @@ impl CheatableSyscallHandler<'_> {
...
+        syscall_handler.syscall_handler.resources,
+        syscall_handler.syscall_handler.context,
+    )?;
+
+    let raw_retdata = &call_info.execution.retdata.0;
+
+    if call_info.execution.failed {
+        // TODO(spapini): Append an error word according to starknet spec if needed.
+        // Something like "EXECUTION_ERROR".
+        return Err(SyscallExecutionError::SyscallError {
```
Comment (piotmag769) [inline crates/cheatable-starknet/src/rpc.rs:622]: Is it copied from cairo? If so can we mark what parts are modified by us? Like this:
https://github.com/foundry-rs/starknet-foundry/pull/309#discussion_r1276194005

File: crates/cheatable-starknet/src/rpc.rs [hunk 4]
```diff
@@ -474,9 +482,76 @@ impl CheatableSyscallHandler<'_> {
...
+                vm,
+                call_contract_syscall,
+                constants::CALL_CONTRACT_GAS_COST,
+            );
+        }
+
         self.syscall_handler.execute_next_syscall(vm, hint)
     }
 
+    fn execute_syscall<Request, Response, ExecuteCallback>(
```
Comment (piotmag769) [inline crates/cheatable-starknet/src/rpc.rs:499]: Could you explain how using it fixes the issue? Is it about `execute_inner_call` and invoking `execute_call_entry_point`?

**Review by piotmag769 — COMMENTED** [review]
declare and deploy changed signatures, please rebase too master (in like 5 minutes or so)

File: crates/forge/tests/integration/start_prank.rs [hunk 5]
```diff
@@ -177,3 +177,69 @@ fn start_prank_in_constructor_test() {
...
+                
+                let class_hash = declare('PrankCheckerProxy');
+                let prepared = PreparedContract { class_hash: class_hash, constructor_calldata: @ArrayTrait::new() };
+                let proxy_contract_address = deploy(prepared).unwrap();
+                let proxy_contract_address: ContractAddress = proxy_contract_address.try_into().unwrap();
+                let proxy_dispatcher = IPrankCheckerProxyDispatcher { contract_address: proxy_contract_address };
+
+                let caller_address = proxy_dispatcher.get_prank_checkers_caller_address(prank_checker_contract_address);
+                assert(caller_address == 234, caller_address);
+            }
```
Comment (piotmag769) [inline crates/forge/tests/integration/start_prank.rs:218]: We should make sure the address changed/block number to the previous one (applies to all prank and roll) 

```
let a = call get_block_number

start_roll 234
let x = call get_block_number
assert x == 234

stop roll
let b = call_get_block_number

assert a == b
```

Comment (MaksymilianDemitraszek) [reply crates/forge/tests/integration/start_prank.rs:218]: https://github.com/foundry-rs/starknet-foundry/issues/341

File: crates/cheatable-starknet/src/rpc.rs (hunk 4 above)
Comment (MaksymilianDemitraszek) [reply crates/cheatable-starknet/src/rpc.rs:499]: Syscall handler when recursively calls contracts. In the previous version we replaced syscall handler only for the top level call and recursive calls have been using original syscall handler without overwritten logic.

Comment (piotmag769) [reply crates/cheatable-starknet/src/rpc.rs:499]: Okay, so I got it right

File: starknet-foundry/crates/cheatable-starknet/src/rpc.rs (hunk 2 above)
Comment (MaksymilianDemitraszek) [reply starknet-foundry/crates/cheatable-starknet/src/rpc.rs]: I haven't investigated that, it is a constraint enforced by blouckifier codebase

File: crates/cheatable-starknet/src/rpc.rs (hunk 3 above)
Comment (MaksymilianDemitraszek) [reply crates/cheatable-starknet/src/rpc.rs:631]: Added some extra documentation

**Review by piotmag769 — APPROVED** [review]

**Review by Radinyn — APPROVED** [review]

File: crates/forge/tests/integration/start_prank.rs (hunk 5 above)
Comment (Radinyn) [reply crates/forge/tests/integration/start_prank.rs]: #303 added some of it

**Review by abulenok — APPROVED** [review]

**Pull request merged**